src/NHibernate.Test/NHSpecificTest/GH1526/Fixture.cs

```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using NHibernate.DomainModel;
+using NHibernate.Linq.Visitors;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH1526
+{
+	[TestFixture]
+	public class Fixture
+	{
+		[Test]
+		public void ShouldCreateDifferentKeyForAnonymousTypesFromDifferentAssemblies()
+		{
+			var exp1 = AnonymousTypeExpressionProviderFromNHibernateTestAssembly.GetExpression();
+			var exp2 = AnonymousTypeExpressionProviderFromNHibernateDomainModelAssembly.GetExpression();
+
+			var key1 = ExpressionKeyVisitor.Visit(exp1, null);
+			var key2 = ExpressionKeyVisitor.Visit(exp2, null);
```

Comment:
Better provide a non-null second argument(`new Dictionary<ConstantExpression, NamedParameter>()`). Although it lacks some pre-condition checks, `ExpressionKeyVisitor` does not actually support receiving it as `null`, and will fail on some expressions if it was `null`.

Reply:
I've followed with your remark in the new commit. However, I believe the real change should be in `ExpressionKeyVisitor` itself, to either a.) make the second arg optional, or b.) default to `new Dictionary<...>` upon receiving `null` in `Visit`/`ctor`.

a.) would be cleaner, although would be a breaking change, since the type is public (does it have to be public, though?)
b.) would be the opposite: non-breaking, but leaving the method's api somewhat dirty

The method is called more times with `null` than with real dictionary, so some kind of cleanup there seems due. Separate PR, though, I reckon.

Reply:
> The method is called more times with null than with real dictionary,

Indeed. Not a good thing since it may then throw a null ref exception.

I am not for b., this just hides the trouble. Well, a. maybe too. I would rather instead, if some change is done (but it is not the subject of this PR), throw an `InvalidOperationException` in `VisitConstant` if `_constantToParameterMap` is `null`.

Reply:
#1536 added for this subject.